atecc608a_se.c

```diff
@@ -0,0 +1,227 @@
+/**
+ * \file atecc608a_se.c
+ * \brief Secure element driver implementation for ATECC508A and ATECC509A.
+ */
+
+/*
+ *  Copyright (C) 2019, ARM Limited, All Rights Reserved
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+#include "atecc608a_se.h"
+#include "atca_helpers.h"
+
+#ifdef DEBUG_PRINT
+#include <stdio.h>
+#endif
+
+#include <stdbool.h>
+#include <stdint.h>
+
+/* Uncomment to print results on success */
+//#define DEBUG_PRINT
+
+/** This macro checks if the result of an `expression` is equal to an
+ *  `expected` value and sets a `status` variable of type `psa_status_t` to
+ *  `PSA_SUCCESS`. If they are not equal, the `status` is set to
+ *  `psa_error instead`, and the code jumps to the `exit` label. */
+#define ASSERT_STATUS(expression, expected, psa_error)          \
+    do                                                          \
+    {                                                           \
+        ATCA_STATUS ASSERT_result = (expression);               \
+        ATCA_STATUS ASSERT_expected = (expected);               \
+        if ((ASSERT_result) != (ASSERT_expected))               \
+        {                                                       \
+            status = (psa_error);                               \
+            goto exit;                                          \
+        }                                                       \
+        status = PSA_SUCCESS;                                   \
+    } while(0)
+
+/** Check if an ATCA operation is successful, translate the error otherwise. */
+#define ASSERT_SUCCESS(expression) ASSERT_STATUS(expression, ATCA_SUCCESS, \
+                                      atecc608a_to_psa_error(ASSERT_result))
+
+ATCAIfaceCfg atca_iface_config = {
```

Comment:
This can be `const`, to same some RAM.

I also see this is public. Why not `static`? What use is it outside the driver? I guess if folks want to use the same configuration for the SE that the driver is using, they could use this. It's only going to be useful for PSA Services and not applications (like our example), though.

Reply:
This cannot be const, because atca_init requires a non-const object, and parts the object itself are probably modified by the implementation (there is, for example an opaque configuration data pointer that we shouldn't interfere with).
This is not static because of the ATCAB_INIT macro - each use of it uses atca_iface_config. ATCAB_INIT is used by our sample app, but I can imagine other code using it too. For non-standard hardware configurations however, I think that we should consider injecting the interface config into the SE configuration.

Reply:
OK, was not aware that this needed to be writable. That's unfortunate.

Reply:
What about provisioning and miscellaneous operations, like config/serial number printing? I'm not sure if we'll ever have a place for that, and still it's useful. I've duplicated the code in my first approach, but Gilles deemed it a bigger issue.

Reply:
It's a big design issue we haven't addressed yet. We do need a way to provision SEs, but how that can be done isn't designed yet.

Reply:
There are two ways to do provisioning. One is to call `psa_import_key`, and it's already fully supported in the interface. The other is that keys are provisioned by bypassing the normal API (typically, they appear “by magic” during manufacturing), and we're still working on how to declare such keys so that they can be used via PSA.